### PR TITLE
WIP: Overload getindex to allow variables to be accessed with []

### PR DESCRIPTION
**Pull request by odow**
@reganbaucke proposed the following as a nicer way to associate variables with models.

It's particularly attractive if you are copying models around and losing track of which scope `x` is in.

### CODE REVIEW

**Comment by mlubin** [comment]
Mmmm... someone's going to come along and ask why this only works for variables and not constraints. This seems like a reasonable direction but we need a more well thought out concept of what's in the "scope" of a model and how ``getvariable`` and ``getconstraint`` relate.

**Comment by odow** [comment]
@mlubin 

Four cases:

1. `name` is a variable but not a constraint in the model. `m[:name]` returns the variable
2. `name` is a constraint but not a variable in the model. `m[:name]` returns the constraint
3. `name` is a variable and a constraint in the model. `m[:name]` errors. Use `getvariable` or `getconstraint` instead
4. `name` is neither a variable nor a constraint in the model. `m[:name]` errors.

Multpiple constraints or variables with the same name are not accessible by their name to avoid ambiguity. Use anonymous variables instead.

**Comment by mlubin** [comment]
Why not just get rid of (deprecate) ``getvariable`` and ``getconstraint`` and have a single dictionary of things that you can stuff into a JuMP model?
Then
```julia
@variable(m, x[1:N])
```
becomes a shorthand for
```julia
m[:x] = @variable(m, [1:N])
```
which seems nice and easy to explain.

**Comment by odow** [comment]
So maybe something like this? I changed the duplicate warning to an error. There is no reason why people shouldn't just use anonymous variables.

**Comment by mlubin** [comment]
> I changed the duplicate warning to an error. There is no reason why people shouldn't just use anonymous variables.

That's fine. The warning was there for deprecation purposes.

**Comment by mlubin** [comment]
I like how this looks. Any reason not to implement ``setindex``?

**Comment by odow** [comment]
You mean so people can do
```julia
m[:x] = @variable(m)
```

Seems reasonable.

**Comment by mlubin** [comment]
Yes. People will misguidedly end up using this syntax to store cat pictures and databases inside the model, but I don't see a reason to get in the way of that.

**Comment by joehuchette** [comment]
I like this, but I think the deprecated ``getvariable`` and ``getconstraint`` will have different behavior in cases like:
```jl
using JuMP
m = Model()
@variable(m, x)
c = getconstraint(m, :x)
```

**Comment by mlubin** [comment]
@joehuchette, so? The main point is to not break existing code that works.

**Comment by joehuchette** [comment]
This existing code breaks:
```jl
using JuMP
m = Model()
@variable(m, x)
@constraint(m, x, x == 1)
```

**Comment by joehuchette** [comment]
My point being that mashing two dictionaries into one can lead to weird conflicts, especially when you're using keys programmatically.

**Comment by odow** [comment]
We could change the error back to a warning and say 

```julia
if haskey(m.objDict, name)
    warn("Overwriting the object $name stored in the model. Consider using anonymous variables and constraints instead")
end
```

**Comment by odow** [comment]
Although I think `@constraint(m, x, x==1)` is far too confusing and should be forbidden syntax.

**Comment by mlubin** [comment]
I'm ok with breaking code that has a variable and a constraint with the same name. Throwing a warning in this case isn't the right answer because we can't guarantee that the behavior will remain the same as before.

**Comment by joehuchette** [comment]
> I'm ok with breaking code that has a variable and a constraint with the same name.
I think this is a lot more common pattern than you're giving credit for (only so many 1-letter variable names...).

**Comment by mlubin** [comment]
> I think this is a lot more common pattern than you're giving credit for (only so many 1-letter variable names...).

It doesn't bother me if it's common since I think this is a change in the right direction for the long term. If we want to maintain separate variables and constraints dictionaries for deprecation purposes that would be ok but not necessary in my view.

**Comment by joehuchette** [comment]
> It doesn't bother me if it's common since I think this is a change in the right direction for the long term.

It's maybe not good practice, but I think name conflicts are inevitable when multiple people/packages are working on the same model, and that there's not much we can do to dissuade people from that.

> If we want to maintain separate variables and constraints dictionaries

I think this is probably the way to go.

**Comment by mlubin** [comment]
> > If we want to maintain separate variables and constraints dictionaries

> I think this is probably the way to go.

This means that we can't throw an error when there's a conflicting name in the new ``objDict``, otherwise there's no point in maintaining separate dictionaries for legacy reasons.

**Comment by joehuchette** [comment]
The behavior of ``[]`` and ``getvariable`` don't have to coincide, I just don't think breaking existing code is a good idea.

**Comment by odow** [comment]
There have been plenty of other times that JuMP changes broke existing code. The solution is pretty simple for current users who might use the functionality
```julia
m=Model()
@variable(m, x)

# change
@constraint(m, x, x<=1)
# to
x = @constraint(m, x<=1)
```

Long-term, I think this is simpler, there is less confusion around scoping, and it means (for better or worse), that people can store data in `m[key]`.

**Comment by joehuchette** [comment]
I agree that we should push users to make this change. Also, to be clear, I'm more concerned about cases like:

```jl
m = Model()
@variable(m, x[1:1_000])
@variable(m, c) # cost, or something
add_other_constraints!(m)

# deep in someone elses code/package...
function add_other_constraints!(model)
    ...
    @constraint(model, c[1:100], ...)
    ...
end
```

I agree that ``@constraint(m, x, x<=1)`` is pathological.

**Comment by mlubin** [comment]
The ``# deep in someone elses code/package...`` code should absolutely not be using named variables.

**Comment by mlubin** [comment]
CC @ccoffrin

**Comment by joehuchette** [comment]
That's not spelled out anywhere in our documentation AFAIK. It's also perfectly valid code, and people are gonna write it.

**Comment by mlubin** [comment]
@joehuchette, if we turn duplicate names into a warning and throw an error only when a duplicately named object is accessed, then it seems like it would avoid breaking most of the casual use cases. It would only break (gracefully, with a useful error) on your example above if someone tried ``getconstraint(model,:c)``.

**Comment by joehuchette** [comment]
What about your concern from before? I think that's better, but I think keeping the old implementation with two dictionaries during the deprecation period is preferable.

File: src/JuMP.jl [hunk 1]
```diff
@@ -804,45 +797,37 @@ end
...
 
-function getconstraint(m::Model, conname::Symbol)
-    if !haskey(m.conDict, conname)
-        error("No constraint with name $conname")
-    elseif m.conDict[conname] === nothing
-        error("Multiple constraints with name $conname")
+# allow easy accessing of JuMP Variables and Constraints
+function Base.getindex(m::JuMP.Model, name::Symbol)
+    if !haskey(m.objDict, name)
+        error("No object with name $name")
```
Comment (mlubin) [inline src/JuMP.jl]: It would be consistent to throw a ``KeyError`` here

**Comment by mlubin** [comment]
> What about your concern from before?

Which concern? There is no possible way to avoid breaking your example if we throw an error on duplicate names.

**Comment by joehuchette** [comment]
Yeah, I guess we could potentially make it throw an error upon trying to use ``[]`` to retrieve a duplicate name, but that's not worth the trouble. The warning is more appealing to me now.

**Comment by mlubin** [comment]
What I'm proposing is a warning when creating duplicate names and an error when retrieving.

**Comment by joehuchette** [comment]
In that case, the error might as well be restricted to ``[]``.

**Comment by mlubin** [comment]
I will personally mail a JuliaOpt sticker to whoever gets hit by an error from
```julia
m = Model()
@variable(m, x[1:1_000])
@variable(m, c) # cost, or something
add_other_constraints!(m)

# deep in someone elses code/package...
function add_other_constraints!(model)
    ...
    @constraint(model, c[1:100], ...)
    ...
end
getconstraint(m, :c)
```

**Comment by joehuchette** [comment]
Famous last words!

**Comment by mlubin** [comment]
It's on the record now :)

**Comment by ccoffrin** [comment]
I am curious about the comment,
```
The # deep in someone else's code/package... code should absolutely not be using named variables.
```
Why not?  And what are the alternatives?

One of the software design challenges I am facing is how two more-or-less independent model building packages can be composed, esp. avoiding variable name clashes in a shared JuMP model.

**Comment by blegat** [comment]
People might also like to use named variables because that gives pretty printing.
```julia
m = Model()
@variable(m, [1:2])
print(m)
```
gives
```
Min 0
Subject to
 __anon__[i] ∀ i ∈ {1,2} -> not as nice as x[i]
```

**Comment by mlubin** [comment]
> One of the software design challenges I am facing is how two more-or-less independent model building packages can be composed, esp. avoiding variable name clashes in a shared JuMP model.

Each model building package should maintain its own "namespace" by storing extra data in the model's extension dictionary or some other mechanism.

> People might also like to use named variables because that gives pretty printing.

We have a separate mechanism to set names with the ``basename`` keyword.

**Comment by joehuchette** [comment]
@ccoffrin one other motivation is that the anonymous syntax is hopefully clearer as to what's going on:  ``x = @variable(m)`` is very clearly an assignment to ``x``, which can be used programmatically like any other variable you might have in your julia code.

**Comment by joehuchette** [comment]
This looks good to me now, thanks @odow!

File: doc/refmodel.rst [hunk 2]
```diff
@@ -57,8 +57,7 @@ Methods
...
 * ``JuMP.build(m::Model)`` - builds the model in memory at the MathProgBase level without optimizing.
 * ``setsolver(m::Model,s::AbstractMathProgSolver)`` - changes the solver which will be used for the next call to ``solve()``, discarding the current internal model if present.
-* ``getvariable(m::Model,name::Symbol)`` - returns the variable or group of variables of the given name which were added to the model with ``@variable``. Throws an error if multiple variables were created with the same name.
-* ``getconstraint(m::Model,name::Symbol)`` - returns the constraint or group of constraint of the given name which were added to the model with ``@constraint`` or ``@NLconstraint``. Throws an error if multiple constraints were created with the same name.
+* ``getindex(m::Model,name::Symbol)`` - returns the variable, or group of variables, or constraint, or group of constraints, of the given name which were added to the model.
```
Comment (mlubin) [inline doc/refmodel.rst]: document ``setindex!`` as well?

**Comment by odow** [comment]
@mlubin you'd better get your stickers ready...

**Pull request merged**